gputils/doc/gpvc.1
.TH gpvc 1 "(c) 2002, 2003 James Bowman, Craig Franklin"
.SH NAME
gpvc - GNU PIC COD file Viewer
.SH SYNOPSIS
.B gpvc [options] file
.SH WARNING
The information in this man page is an extract from the full documentation of
gputils and is limited to the meaning of the options.  For complete and 
current documentation, refer to "gputils.ps" located in the gputils docs 
directory.
.SH DESCRIPTION
.B gpvc
is an COD file viewer for Microchip (TM) PIC (TM) micro-controllers.  COD files
are and output from gpasm.  They contain information used for simulation.
.B gpvc
is part of gputils.  Check the gputils(1) manpage for details on other GNU 
PIC utilities.
.SH OPTIONS
.TP
.B -a, --all
Display all information in the COD file
.TP
.B -d, --directory
Display directory header.
.TP
.B -l, --listing
Display source listing cross.
.TP
.B -m, --message
Display debug message area.
.TP
.B -h, --help
Show the usage message. 
.TP
.B -r, --rom 
Display rom.
.TP
.B -s, --symbols
Display symbols.
.TP
.B -v, --version
Show the version.
.SH SEE ALSO
gputils(1)
.SH AUTHOR
Craig Franklin <craigfranklin@attbi.com>
.SH COPYRIGHT
Copyright (C) 2002, 2003 James Bowman, Craig Franklin

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


gputils/ChangeLog
2014-01-25  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * doc/gputils.lyx, header/Makefile.am, header/Makefile.in,
    header/p12f1612.inc, header/p12lf1612.inc, header/p16f1613.inc,
    header/p16lf1613.inc, header/p16f1xxx.inc, libgputils/gpcfg-table.c,
    libgputils/gpprocessor.c, lkr/Makefile.am, lkr/Makefile.in,
    lkr/12f1612_g.lkr, lkr/16f1613_g.lkr, lkr/12lf1612_g.lkr,
    lkr/16lf1613_g.lkr:
    New devices based on the mplabx-2.0.

  * header/p12f1571.inc, header/p12f1572.inc, header/p12lf1571.inc,
    header/p12lf1572.inc, header/p16f1703.inc, header/p16f1707.inc,
    header/p16f1713.inc, header/p16f1716.inc, header/p16f1782.inc,
    header/p16f1783.inc, header/p16f1784.inc, header/p16f1786.inc,
    header/p16f1787.inc, header/p16f1788.inc, header/p16f1789.inc,
    header/p16lf1703.inc, header/p16lf1707.inc, header/p16lf1713.inc,
    header/p16lf1716.inc, header/p16lf1782.inc, header/p16lf1783.inc,
    header/p16lf1784.inc, header/p16lf1786.inc, header/p16lf1787.inc,
    header/p16lf1788.inc, header/p16lf1789.inc, lkr/*_g.lkr:
    Updated based on the mplabx-2.0.

2014-01-22  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * header/p12f1822.inc, header/p12f1840.inc, header/p12lf1822.inc,
    header/p12lf1840.inc, header/p12lf1840t39a.inc, header/p12lf1840t48a.inc:
    Added the missing CONFIG2 Options: _DEBUG_ON, _DEBUG_OFF
    "#62 Missing CONFIG2 DEBUG file definition in some pic12f inc files"
    thanks to Ludrao.

2013-11-09  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * doc/gputils.lyx, header/Makefile.am, header/Makefile.in,
    header/p16f1703.inc, header/p16f1705.inc, header/p16f1707.inc,
    header/p16f1709.inc, header/p16lf1703.inc, header/p16lf1705.inc,
    header/p16lf1707.inc, header/p16lf1709.inc,
    libgputils/gpcfg-table.c, libgputils/gpprocessor.c,
    lkr/16f1703_g.lkr, lkr/16f1705_g.lkr, lkr/16f1707_g.lkr,
    lkr/16f1709_g.lkr, lkr/16lf1703_g.lkr, lkr/16lf1705_g.lkr,
    lkr/16lf1707_g.lkr, lkr/16lf1709_g.lkr, 
    lkr/Makefile.am, lkr/Makefile.in:
    New devices based on the mplabx-1.95.

  * header/p12f1571.inc, header/p12f1572.inc, header/p12f1822.inc,
    header/p12lf1571.inc, header/p12lf1572.inc, header/p16f1784.inc,
    header/p16lf1784.inc,
    lkr/12f1572_g.lkr, lkr/12lf1572_g.lkr, lkr/16f1509_g.lkr,
    lkr/16f1783_g.lkr, lkr/16f1936_g.lkr, lkr/16f1937_g.lkr,
    lkr/16lf1509_g.lkr, lkr/16lf1783_g.lkr, lkr/16lf1906_g.lkr,
    lkr/16lf1907_g.lkr, lkr/16lf1936_g.lkr, lkr/16lf1937_g.lkr:
    Updated based on the mplabx-1.95.

  * AUTHORS:
    Corrected my email.

2013-10-12  Borut Razem  <borutr@users.sourceforge.net>
  * libgputils/gpmemory.c: fixed b_memory_get()

2013-10-11  Borut Razem  <borutr@users.sourceforge.net>
  * libgputils/gpmemory.c: fixed bug
    #269 18F26j50, gplink crash on WPDIS

2013-09-25  Borut Razem  <borutr@users.sourceforge.net>
  * gpasm/scan.l, gpasm/ppscan.l: fixed bug
    #270 Gputils doesn't compile with newer versions of bison and flex

2013-09-08  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * doc/gputils.lyx, header/Makefile.am, header/Makefile.in,
    header/p12f1571.inc, header/p12f1572.inc, header/p12lf1571.inc,
    header/p12lf1572.inc, header/p16f1713.inc, header/p16f1716.inc,
    header/p16f1829lin.inc, header/p16lf1713.inc, header/p16lf1716.inc,
    libgputils/gpcfg-table.c, libgputils/gpprocessor.c,
    lkr/12f1571_g.lkr, lkr/12f1572_g.lkr, lkr/12lf1571_g.lkr,
    lkr/12lf1572_g.lkr, lkr/16f1713_g.lkr, lkr/16f1716_g.lkr,
    lkr/16f1829lin_g.lkr, lkr/16lf1713_g.lkr, lkr/16lf1716_g.lkr,
    lkr/Makefile.am, lkr/Makefile.in:
    New devices based on the mplabx-1.9.

2013-09-08  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * header/*.inc:
    Updated based on the mplabx-1.9.

2013-08-09  Borut Razem  <borutr@users.sourceforge.net>
  * config.guess, config.sub: updated from git head
  * depcomp: updated
  * README: updated MinGW build info

2013-08-05  Borut Razem  <borutr@users.sourceforge.net>
  * gplink/gplink.c, libgputils/gpcofflink.c, libgputils/gpcofflink.h,
    gplink/gplink.h.in, doc/gputils.lyx, man/gplink.1.in,
    man/fr/gplink.1.in: implemented --mplink-compatble in gplink
  * gplink/testsuite/gplink.project/asmfiles/op14e.asm,
    gplink/testsuite/gplink.project/lkrfiles/op14e.lkr,
    gplink/testsuite/gplink.project/hexfiles/op14e.hex:
    added gplink regression tests for bug
    #268 PIC14e: HIGH directive not setting bit-7

2013-08-02  Borut Razem  <borutr@users.sourceforge.net>
  * gpasm/testsuite/gpasm.project/asmfiles/op14e.asm,
    gpasm/testsuite/gpasm.project/listfiles/op14e.lst,
    gpasm/testsuite/gpasm.project/hexfiles/op14e.hex,
    gpasm/testsuite/gpasm.project/objasm/op14e.asm,
    gpasm/testsuite/gpasm.project/objlstfiles/op14e.lst,
    gpasm/testsuite/gpasm.project/objfiles/op14e.o:
    added gpasm regression tests for bug
    #268 PIC14e: HIGH directive not setting bit-7

2013-08-01  Borut Razem  <borutr@users.sourceforge.net>
  * gpasm/evaluate.c, libgputils/gpcofflink.c, libgputils/gpprocessor.c,
    libgputils/gpprocessor.h: fixed bug
    #268 PIC14e: HIGH directive not setting bit-7
  * gputils/gpvo.c: display symbolic value of relocation type

2013-07-24  Borut Razem  <borutr@users.sourceforge.net>
  * libgputils/gpcofflink.c, libgputils/gpprocessor.c:
    applied modified patch [PATCH] Bad warning for BRA on 16F code,
    thanks to Richard Hodges

2013-06-30  Molnar Karoly  <molnarkaroly@users.sourceforge.net>
  * header/p16f753.inc, p16hv753.inc, header/p16f1704.inc,
    header/p16f1708.inc, header/p16lf1704.inc, header/p16lf1708.inc,
    lkr/16f753_g.lkr, lkr/16hv753_g.lkr:
    Fresh based on the mplabx-1.85.

2013-05-17  Borut Razem  <borutr@users.sourceforge.net>
  * gpasm/directive.c: generate error if CONFIG directive is used with
    non 16 bit devices

2013-05-14  Borut Razem  <borutr@users.sourceforge.net>
  * doc/html-help/Makefile.am, doc/html-help/Makefile.in:

trunk/gputils/gputils/gpvc.c
/* Displays contents of ".COD" files
   Copyright (C) 2001 Scott Dattalo

This file is part of gputils.
 
gputils is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.
 
gputils is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
 
You should have received a copy of the GNU General Public License
along with gputils; see the file COPYING.  If not, write to
the Free Software Foundation, 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.  */

#include "stdhdr.h"

#include "gpvc.h"
#include "dump.h"
#include "block.h"

FILE *codefile;  
char filename[BUFFER_LENGTH];
char temp[BLOCK_SIZE];
char *source_file_names[MAX_SOURCE_FILES];
FILE *source_files[MAX_SOURCE_FILES];
DirBlockInfo main_dir;

int addrsize;

char directory_block_data[BLOCK_SIZE];
char * SymbolType4[154] = {
  "a_reg          ", "x_reg          ", "c_short        ", "c_long         ",
  "c_ushort       ", "c_ulong        ", "c_pointer      ", "c_upointer     ",
  "table          ", "m_byte         ", "m_boolean      ", "m_index        ",
  "byte_array     ", "u_byte_array   ", "word_array     ", "u_word_array   ",
  "func_void_none ", "func_void_byte ", "funcVoidTwobyte", "func_void_long ",
  "func_void_undef", "func_int_none  ", "func_int_byte  ", "func_int_twobyt",
  "func_int_long  ", "func_int_undef ", "func_long_none ", "func_long_byte ",
  "funcLongTwobyte", "func_long_long ", "func_long_undef", "pfun_void_none ",
  "pfun_void_byte ", "pfunVoidTwobyte", "pfun_void_long ", "pfun_void_undef",
  "pfun_int_none  ", "pfun_int_byte  ", "pfunIntTwobyte ", "pfun_int_long  ",
  "pfun_int_undef ", "pfun_long_none ", "pfun_long_byte ", "pfun_long_twoby",
  "pfun_long_long ", "pfun_long_undef", "address        ", "constant       ",
  "bad_und        ", "br_und         ", "upper_und      ", "byte_und       ",
  "add_und        ", "m_keyword      ", "add_mi_und     ", "vector         ",
  "port_w         ", "port_r         ", "port_rw        ", "port_rmw       ",
  "endif          ", "exundef        ", "macro_t        ", "macro_s        ",
  "macro_a        ", "macro_c        ", "c_keyword      ", "void           ",
  "c_enum         ", "c_typedef1     ", "c_utypedef1    ", "c_typedef2     ",
  "c_utypedef2    ", "cp_typedef1    ", "cp_utypedef1   ", "cp_typedef2    ",
  "cp_utypedef2   ", "c_struct       ", "i_struct       ", "l_const        ",
  "s_short        ", "s_ushort       ", "s_long         ", "s_ulong        ",
  "sa_short       ", "sa_ushort      ", "sa_long        ", "sa_ulong       ",
  "sp_short       ", "sp_ushort      ", "sp_long        ", "sp_ulong       ",
  "FixedPointer   ", "PointerFunction", "cc_reg         ", "PtrF_void_none ",
  "PtrF_void_byte ", "PtrF_void_twobyt", "PtrF_void_long ", "PtrF_void_undef",
  "PtrF_int_none  ", "PtrF_int_byte  ", "PtrF_int_twobyte", "PtrF_int_long  ",
  "PtrF_int_undef ", "PtrF_long_none ", "PtrF_long_byte ",
  "PtrF_long_twobyte", "PtrF_long_long ", "PtrF_long_undef",
  "PfAR_void_none ", "PfAR_void_byte ", "PfAR_void_twobyte",
  "PfAR_void_long ", "PfAR_void_undef", "PfAR_int_none  ", "PfAR_int_byte  ",
  "PfAR_int_twobyte", "PfAR_int_long  ", "PfAR_int_undef ", "PfAR_long_none ",
  "PfAR_long_byte ", "PfAR_long_twobyte", "PfAR_long_long ",
  "PfAR_long_undef", "FWDlit         ", "Pfunc          ", "mgoto          ",
  "mcgoto         ", "mcgoto2        ", "mcgoto3        ", "mcgoto4        ",
  "mcgoto74       ", "mcgoto17       ", "mccall17       ", "mcall          ",
  "mc_call        ", "res_word       ", "local          ", "unknown        ",
  "varlabel       ", "external       ", "global         ", "segment        ",
  "Bankaddr       ", "bit_0          ", "bit_1          ", "bit_2          ",
  "bit_3          ", "bit_4          ", "bit_5          ", "bit_6          ",
  "bit_7          ", "debug          "
};


void show_usage(void)
{
  printf("Usage: gpdasm <options> <filename>\n");
  printf("Where <options> are:\n");
  printf("  -a, --all          Display all information in .cod file\n");
  printf("  -d, --directory    Display directory header\n");
  printf("  -l, --listing      Display source listing cross\n");
  printf("  -m, --message      Display debug message area\n");
  printf("  -h, --help         Show this usage message \n");
  printf("  -r, --rom          Display rom\n");
  printf("  -s, --symbols      Display symbols\n");
  printf("  -v, --version      Show version\n");
  printf("\n");
  printf("Report bugs to:\n");
  printf("%s\n", BUG_REPORT_URL);
  exit(0);
}

#define GET_OPTIONS "?adhlmrsv"

  /* Used: adhlmrsv */
  static struct option longopts[] =
  {
    { "all",         0, 0, 'a' },
    { "directory",   0, 0, 'd' },
    { "help",        0, 0, 'h' },
    { "listing",     0, 0, 'l' },
    { "message",     0, 0, 'm' },
    { "rom",         0, 0, 'r' },
    { "symbols",     0, 0, 's' },
    { "version",     0, 0, 'v' },
    { 0, 0, 0, 0 }
  };

#define GETOPT_FUNC getopt_long(argc, argv, GET_OPTIONS, longopts, 0)

int main(int argc, char *argv[])
{
  extern int optind;
  unsigned int buffer_size;
  int c,usage=0;
  int display_flags;
  Directory *dir;

  addrsize = 1;

#define DISPLAY_NOTHING 0
#define DISPLAY_DIR     1
#define DISPLAY_SYM     2
#define DISPLAY_ROM     4
#define DISPLAY_SRC     8
#define DISPLAY_MESS    16
#define DISPLAY_ALL     0xff

  display_flags = DISPLAY_NOTHING;
  while ((c = GETOPT_FUNC) != EOF) {
    switch (c) {
    case '?':
    case 'h':
      usage = 1;
      break;
    case 'a':
      display_flags = DISPLAY_ALL;
      break;
    case 'd':
      display_flags |= DISPLAY_DIR;
      break;
    case 's':
      display_flags |= DISPLAY_SYM;
      break;
    case 'r':
      display_flags |= DISPLAY_ROM;
      break;
    case 'l':
      display_flags |= DISPLAY_SRC;
      break;
    case 'm':
      display_flags |= DISPLAY_MESS;
      break;
    case 'v':
      fprintf(stderr, "%s\n", GPVC_VERSION_STRING);
      exit(0);
    }
    if (usage)
      break;
  }
  
  if (optind < argc)
    strcpy(filename, argv[optind]);
  else
    usage = 1;

  if(display_flags == DISPLAY_NOTHING)
    display_flags = DISPLAY_ALL;


  if (usage) {
    show_usage();
  }

  codefile = fopen(filename,"r");
  if(codefile == NULL) {
    printf("Unable to open %s\n",filename);
    exit(1);
  }


  /* Start off by reading the directory block */
  read_directory();

  fseek(codefile, 0,SEEK_SET);
  buffer_size = fread(directory_block_data, 1, BLOCK_SIZE, codefile);

  if(display_flags & DISPLAY_DIR)
    directory_block();

  dir = (Directory *)directory_block_data;

  if(display_flags & DISPLAY_ROM)
    dump_code();

  if(display_flags & DISPLAY_SYM) {
    dump_symbols();
    dump_lsymbols();
    dump_local_vars();
  }

  dump_source_files();

  if(display_flags & DISPLAY_SRC)
    dump_line_symbols();

  if(display_flags & DISPLAY_MESS)
    dump_message_area();

  return(0);
}
